### PR TITLE
Remove buffer -> dtype dependency

### PR DESCRIPTION
**Pull request by gatesn**
Part of my refactor to turn vortex-buffer into our equivalent of the "bytes" crate

TODO: remove match_integers_by_width which previously was used for unchecked transmute

### CODE REVIEW

File: vortex-buffer/src/lib.rs [hunk 1]
```diff
@@ -42,31 +41,10 @@ impl Buffer {
...
-            Buffer::Arrow(buffer) => match_each_native_ptype!(T::PTYPE, |$T| {
-                buffer
-                    .into_vec()
-                    .map(|vec| unsafe { std::mem::transmute::<Vec<$T>, Vec<T>>(vec) })
-                    .map_err(Buffer::Arrow)
-            }),
-            // Cannot always convert bytes into a mutable vec
+            Buffer::Arrow(buffer) => buffer.into_vec::<u8>().map_err(Buffer::Arrow),
+            // Cannot convert bytes into a mutable vec
             Buffer::Bytes(_) => Err(self),
```
Comment (robert3005) [inline vortex-buffer/src/lib.rs:48]: why is this an Err? If you are the only one holding the bytes then this should be possible

Comment (gatesn) [reply vortex-buffer/src/lib.rs:48]: I don't think you can go from Bytes back to BytesMut

Comment (robert3005) [reply vortex-buffer/src/lib.rs:48]: https://rustasync.github.io/runtime/bytes/struct.Bytes.html#method.try_mut seems to be the right signature and then you can use that to construct vec from parts?

Comment (gatesn) [reply vortex-buffer/src/lib.rs:48]: 🤦 I missed that. I thought it seemed odd

Comment (gatesn) [reply vortex-buffer/src/lib.rs:48]: Hmmm no, I'm not sure what docs those are but they're not up-to-date for the bytes crate. See also https://github.com/tokio-rs/bytes/issues/350

Comment (robert3005) [reply vortex-buffer/src/lib.rs:48]: Looks like someone is hosting their own fork or some outdated link that google served. You’re right. Maybe we need our own version as well?

Comment (gatesn) [reply vortex-buffer/src/lib.rs:48]: We do anyway to have more control over alignment. This crate tries to abstract that away for now.

**Pull request merged**